### ISSUE TITLE
Not sending transition events.

### ISSUE THREAD

**Issue by guillebot** [owntracks/android#175]
The app isn't sending all the transition events.

Sometimes it sends them, but a lot of times wont. Only when I open the app and navigate it, then send some events.

I'm now subscribing to the beta channel in order to test it.

**Comment by watou** [comment]
I also see that many waypoint transitions are not being sent. I enter and leave waypoints but sometimes only location is sent (without "desc" and "event" JSON keys).

OwnTracks 0.5.11, Nexus 5, Android 5.1.1, mosquitto 1.4.2 on RPi B+, Debian wheezy, subscribed by openHAB 1.7.1.

Reporting:
 Automation location reporting: ON
 Include battery: ON
Notifications:
 All but vibrate are ON
Advanced:
 Topic: default (owntracks/myname/nexus5)
 QOS: 1
 Publish retained: ON
 Subscribe: ON
 Subscribe topic: owntracks/+/+
Remote commands: all OFF
Locator:
 Locator displacement: 500
 Locator interval: 15
 Locator background accuracy: Balanced power
 Locator foreground accuracy: High power
 Fastest reporting interval: 5
Beacons:
 Ranging: OFF
Autostart: OFF

**Comment by binarybucks** [comment]
Please try the latest beta from https://betas.to/bpNWWMqz and see if it works better.

**Comment by watou** [comment]
So far 0.5.20 seems better, but I will take a tour of some waypoints today to verify.  Thank you very much!

**Comment by haunma** [comment]
I am having the same problem with 0.5.22 on a Nexus 5 with Android 4.4.4.  Three waypoints are defined: two around my home (one with radius 1.5 km, the other 15 km), and one around work.  Broker is mosquitto 1.3.4 on debian.  I am connected with TLS and username/password auth, dumping the messages to a log file with mosquitto_sub.  All waypoints are shared, and I get mqtt messages in the log when any of their parameters are changed.  Other app settings:

Automatic location ON
Include battery OFF
Notifications: all but vibrate
Publish QOS 1
Publish retained ON
Messaging ON
Remote report OFF
Remote configuration OFF
Foreground accuracy High Power
Background accuracy Low Power
Background displacement 500m
Background interval 5 minutes
Autostart ON

(@watou has more options than I do, e.g. my version does not have the "Topic" or "Subscribe topic" settings.  Maybe those were removed since 0.5.11?)

In the app, all three waypoints insist "Last transition: never" and sure enough, the mqtt log has no messages of the transition type.  The log does show regular location updates, which move in and out of the circular waypoint regions.  Work is about 20km from home, and I've gone back and forth twice now.  Multiple location updates crossing the geofences, but none of them fire.

I do have fairly restrictive location settings on the phone: battery saver only, with wifi and GPS turned off most of the time.  The mqtt messages are claiming about 1 km accuracy.  That's why I am using such big waypoint radii, in case the app were being picky.  But it did not help.

I have noticed two other bugs:
1) There may be many location updates, e.g. while I am riding the bus home, but the notification bar never changes, even while the phone is awake for a long time, and it still shows my work location which I left a long time ago (30-60 minutes).  When I do a manual report, then the notification bar updates.  Or maybe it is simply giving the Owntracks app focus that does it?  In any case, Owntracks knows where it is, and is sending location updates to the server, but it does not geocode them into addresses and update the bar until something "kicks" it, even with the phone awake the whole time.
2) When I request a manual update, the location is published twice, once with "t":"u" and once without (same lat, lon, and timestamp).

If you agree one or both of these are bugs, I can enter new issues for them.

**Comment by binarybucks** [comment]
I'm currently investigating both issues because I'm too experiencing these.

The notification issue is just a cosmetic one and the code for is being
rewritten anyway.

To test things, could you please keep the battery saver disabled and wifi
enabled? The Google location APIs are fragile enough even without such
influences. Especially geofencing seems to rely on wifi positioning quite a
bit from my testing.

Mit freundlichen Grüßen / With kind regards
Alexander Rust

On 26 Sep 2015, at 02:42, Mark <notifications@github.com> wrote:

I am having the same problem with 0.5.22 on a Nexus 5 with Android 4.4.4.
Three waypoints are defined: two around my home (one with radius 1.5 km,
the other 15 km), and one around work. Broker is mosquitto 1.3.4 on debian.
I am connected with TLS and username/password auth, dumping the messages to
a log file with mosquitto_sub. All waypoints are shared, and I get mqtt
messages in the log when any of their parameters are changed. Other app
settings:

Automatic location ON
Include battery OFF
Notifications: all but vibrate
Publish QOS 1
Publish retained ON
Messaging ON
Remote report OFF
Remote configuration OFF
Foreground accuracy High Power
Background accuracy Low Power
Background displacement 500m
Background interval 5 minutes
Autostart ON

(@watou <https://github.com/watou> has more options than I do, e.g. my
version does not have the "Topic" or "Subscribe topic" settings. Maybe
those were removed since 0.5.11?)

In the app, all three waypoints insist "Last transition: never" and sure
enough, the mqtt log has no messages of the transition type. The log does
show regular location updates, which move in and out of the circular
waypoint regions. Work is about 20km from home, and I've gone back and
forth twice now. Multiple location updates crossing the geofences, but none
of them fire.

I do have fairly restrictive location settings on the phone: battery saver
only, with wifi and GPS turned off most of the time. The mqtt messages are
claiming about 1 km accuracy. That's why I am using such big waypoint
radii, in case the app were being picky. But it did not help.

I have noticed two other bugs:
1) There may be many location updates, e.g. while I am riding the bus home,
but the notification bar never changes, even while the phone is awake for a
long time, and it still shows my work location which I left a long time ago
(30-60 minutes). When I do a manual report, then the notification bar
updates. Or maybe it is simply giving the Owntracks app focus that does it?
In any case, Owntracks knows where it is, and is sending location updates
to the server, but it does not geocode them into addresses and update the
bar until something "kicks" it, even with the phone awake the whole time.
2) When I request a manual update, the location is published twice, once
with "t":"u" and once without (same lat, lon, and timestamp).

If you agree one or both of these are bugs, I can enter new issues for them.

—
Reply to this email directly or view it on GitHub
<https://github.com/owntracks/android/issues/175#issuecomment-143384685>.

**Comment by haunma** [comment]
Hi Alex,

I ran this weekend with location mode set to "high accuracy" and wifi always on.  But still no transitions (there should have been at least half a dozen).

I noticed that the accuracy numbers in the log did not change, despite my location being much more accurate in, e.g. Maps (although I am not seeing a shaded circle there, just a very accurate grey dot).  "acc" field still ranges from ~ 600m to 5km (!), usually around 1-1.5km.

I sympathize re: google APIs.  If there were some way to do this completely outside Play Services, that would be ideal.  One reason I run with the google-wifi-location stuff disabled is that Play Services is a frequent cause of constant wakelocks and battery drain.  I saw there was another issue in the tracker about getting rid of Play Services.  Maybe if fixing this takes more than X amount of effort, the synergy of getting rid of P.S. at the same time will make it worthwhile :|

Mark

**Comment by binarybucks** [comment]
I cannot say yet if it's a Play Services issue. So far the geofence API has been the most reliable (in contrast to the location API), so I guess I screwed up something when registering the Waypoints. 

Looking at the location in maps is not really representative by the way. Google maps does not seem to hook into the Play Services location APIs as other apps ;) 

One thing to keep in mind: Play Services are seen as a frequent cause for wakelocks and battery drain, but just because they receive much power blame for the apps that are using it. Without them, you would perceive less usage because blame would be shared between all the apps using e.g. location services. But as these could not reuse each others locations, the actual energy usage would probably be higher because the phone will spend more time being awake to do redundant work. Play Services might be considered evil, but the reason for this is mostly FUD.

**Comment by haunma** [comment]
Let me know if you'd like help with further testing.  I signed up for the beta access but haven't installed anything yet.

**Comment by haunma** [comment]
Just an update:  I upgraded to client version 0.5.23 and am still seeing the same problem.  (The phone is also set for high-accuracy location mode with always-on wifi.)  Not a single transition into or out of a waypoint has been logged.  I have noticed on a couple occasions, that owntracks is keeping my phone continuously awake, draining the battery in ~ 12 hours.  I can't say for sure if this is new with 0.5.23, and of course it may not be owntracks' "fault"---I am just going by the battery stats.

**Comment by glynhudson** [comment]
I am also running 0.5.23 and have the same issue that no transition events
are reported via mqtt. The transitions in and out of way points are
registered on the phone but not posted to mqtt topic.

- sent from my mobile device

**Comment by haunma** [comment]
This might be a different issue, as in my case the transition events never occur [on the phone].  Of course, the reports of intermittent behavior would suggest that there is > 1 root cause at work here.

**Comment by binarybucks** [comment]
@glynhudson If the waypoint list shows that the waypoint was actually triggered (e.g. 2 hours ago) but you're not seeing any messages on the broker, it can have several reasons that are totally independent from the ones that @haunma is reporting. In your case it could be a configuration issue so that e.g. you cannot publish to this topic. Which mode are you using for this and how do you determine that no messages arrive on the broker? Alternatively, it can also be a code issue because offline queuing currently only works in ~50% of all cases due to a race condition. I'm working on a fix for this. In any way, it could help if you could attach it to Android Studio to collect some LogCat output when a geofence is registered and triggered. Please drop of a note to support[at]owntracks.org if you would like me to assist you with this. 

@haunma In your case, something seems to be severely wrong. I haven't seen such a behavior on any other device yet. Could you check out the following points please? 
1) How do you determine that no transition is triggered? Likewise with the last trigger date in the list or via MQTT? 
2) Do you have the ongoing notification enabled? 
3) Which OS version are you using? 
4) How did you create the waypoints that do not trigger? Manually in the app, via remote config or with an imported configuration file? 
5) How did you start the app? Manually or via the autostart?
7) When you say you are running location mode in "high accuracy", do you mean the Android location settings or the apps background location accuracy? 
7) Can you check if a newly created waypoint at your current location with radius 100m does trigger when you are connected to the broker? 
8) In case your device is rooted, could you install a wakelock detector and check if there are any OwnTracks wakelocks? 
9) When you say  "I have noticed on a couple occasions, that owntracks is keeping my phone continuously awake, draining the battery in ~ 12 hours" how did you determine that?

**Comment by haunma** [comment]
@binarybucks, here are at least partial answers.  Battery life seems better today, but I will install a wakelock detector if I see abnormal battery drain again.

1) All three of my waypoints on the Waypoints screen say "Last transition: never".  Also, no waypoint notifications received on my home mqtt server.
2) Yes, all four options are set ON in the Notifications screen.
3) My phone is a Nexus 5 with Android 4.4.4.  The ROM (Purity) is not stock, but I haven't been able to trace problems with any other app to the ROM, so I doubt this is the problem.
4) Waypoints were created manually in the app.
5) Both, I guess.  It auto-starts on every boot, but I have also killed and restarted it manually when my server IP address changed.
6) Re: high accuracy, I mean the Android location settings are set to maximum privacy-giveaway mode :)
7) I'll give it a try and let you know.  Should it trigger right away, or do I first need to move out of the radius and then back in?
8) Device is rooted; will try this if I noticed battery issues again.
9) This was determined only in a crude way, using the stock battery-stats screen and noticing that Owntracks was [claimed] using ~ 50% of the current charge, more than screen, idle, etc., when the phone was mostly supposed to be asleep.

**Comment by haunma** [comment]
@binarybucks, I got a transition!  Immediately after creating a new waypoint at the current location.

Is it possible that manually entered lat/lon does not work?  (I did double-check the other waypoints BTW, to make sure I didn't have a sign error.)  The only other explanation that comes to mind, is that somehow when a waypoint is created, if the phone is in some weird state, then that waypoint will forever be non-functional, even after the phone is in a normal state.  E.g. when I created those old waypoints I did not have wifi or high-accuracy location enabled.  (And if I have anything to say about it, those will be disabled again shortly :)

Also, is there any way to delete a waypoint?  I wanted to delete my old points, but there is no long-press menu.
